CASSANDRA-16841 Fix unexpectedly ignored dtests

k-rus opened the pull request:
This commit fixes few unexpected behaviors, which were:
1. If any test class is marked with `resource_intensive`, then all tests
   in the file were treated as resource intensive.
2. If one test class is marked with `vnodes` and another test class is
   marked with `no_vnodes` in the same file, then all tests were
   skipped in the file.

The unexpected behaviors are due to marking all tests in a file if any 
class is assigned the mark. This commit fixes that marks on other 
classes in the file do not affect a test, which is not marked and 
neither its class. This is also applied to upgrade tests, so the code 
is simplified and behavior is unified.

Since the upgrade tests were utilizing that the marking one class is 
propagated to all classes in the file, they were revisited. In many 
cases the upgrade tests were already included, because their classes 
inherited from a class, which already was marked. In such cases 
unnecessary marks were removed.

The unit tests of `is_shippable` from `conftest.py` are changed to test
calculating skipping conditions from provided mocked configurations
instead of shipping if conditions are satisfied or not, which required
to change the underlying implementation. The unit tests are also
rewritten to be parameterized and it covers now all possible meaningful
test cases.

patch by Ruslan Fomkin; reviewed by Andrés de la Peña and Ekaterina 
Dimitrova for CASSANDRA-16841

k-rus commented:
[CircleCI build](https://app.circleci.com/pipelines/github/k-rus/cassandra?branch=CASS-16841-CI) is running

adelapena committed:
Place is_skippable inside SkipConditions

adelapena commented on conftest.py:
```diff
@@ -545,53 +544,46 @@ def cassandra_dir_and_version(config):
-        else:
-            logger.info("SKIP: Skipping %s because it is marked with %s" % (item, mark))
+        if skip_marked:
+            logger.info(
+                "SKIP: Skipping %s because it is marked with %s" % (item, mark))
```
Nit: I think we don't need to break this line nor the other `loggger.info` call right below
```suggestion
            logger.info("SKIP: Skipping %s because it is marked with %s" % (item, mark))
```

k-rus replied:
It is not required. I guess I had default setting for autopep8. I will restore it back.

adelapena commented on conftest.py:
```diff
@@ -545,53 +544,46 @@ def cassandra_dir_and_version(config):
+            or _is_skippable(item, "vnodes", skip_cond.skip_vnodes_tests, skip_non_marked=False)
+            or _is_skippable(item, "no_vnodes", skip_cond.skip_no_vnodes_tests, skip_non_marked=False)
+            or _is_skippable(item, "no_offheap_memtables", skip_cond.skip_no_offheap_memtables_tests, skip_non_marked=False)
+            or _is_skippable(item, "depends_driver", skip_marked=True, skip_non_marked=False))
```
Nit: maybe this block would be a bit easier to read if we always name the `skip_marked`/`skip_non_marked` parameters:
```suggestion
    return (self._is_skippable(item, "upgrade_test",
                               skip_marked=self.skip_upgrade_tests,
                               skip_non_marked=self.skip_non_upgrade_tests)
            or self._is_skippable(item, "resource_intensive",
                                  skip_marked=self.skip_resource_intensive_tests,
                                  skip_non_marked=self.skip_non_resource_intensive_tests)
            or self._is_skippable(item, "vnodes",
                                  skip_marked=self.skip_vnodes_tests,
                                  skip_non_marked=False)
            or self._is_skippable(item, "no_vnodes",
                                  skip_marked=self.skip_no_vnodes_tests,
                                  skip_non_marked=False)
            or self._is_skippable(item, "no_offheap_memtables",
                                  skip_marked=self.skip_no_offheap_memtables_tests,
                                  skip_non_marked=False)
            or self._is_skippable(item, "depends_driver",
                                  skip_marked=True,
                                  skip_non_marked=False))
```

adelapena commented on conftest.py:
```diff
@@ -545,53 +544,46 @@ def cassandra_dir_and_version(config):
-                 include_no_offheap_memtables_tests):
-
-    skippable = False
+class SkipConditions:
```
If we are going to pack these vars into a new class perhaps we could also place `_is_skippable` and `is_skippable` inside the class, for example [this way](https://github.com/adelapena/cassandra-dtest/commit/bb3b4382a59c453ccfcd8b85c5cbc70afe8b5cee).

k-rus replied:
Thank you for the suggestion with the patch. Looks nice. I plan to incorporate it into the patch.

adelapena commented on upgrade_tests/paging_test.py:
```diff
@@ -188,6 +188,7 @@ def random_txt(text):
             assert_lists_equal_ignoring_order(pf.all_data(), expected_data, sort_key='value')
 
 
+@pytest.mark.upgrade_test
```
Can we simply add the annotation to `BasePagingTester`, which is the parent of all the marked tests, [this way](https://github.com/adelapena/cassandra-dtest/commit/c3668b62654b0b5f11eccb9e2cfd74d1e164d990)?

k-rus replied:
Thank you for pointing it. Actually marking `BasePagingTester` is not needed, since it inherits form `UpgradeTester`, which is already marked. My main concern is that the marking becomes implicit. I don't know if it will be confusing or not, while  believe developers will have intention that all tests are marked as upgrade tests in this folder. I can try to clarify in `upgrade_tests/README.md` that using `UpgradeTester` is enough.

I tested that it works in `pytest`, however the readme instructions use `nosetests` to run upgrade tests.

adelapena replied:
I think that the note in the readme makes this clearer, thanks. At some point we will have to clear the references to nosetests, but I think there is no need to do it during this ticket.

adelapena commented on .gitignore:
```diff
@@ -10,3 +10,4 @@ last_test_dir
 upgrade
 html/
 doxygen/doxypy-0.4.2/
+.pytest_cache/
```
+1000

adelapena commented on conftest.py:
```diff
@@ -605,28 +597,17 @@ def pytest_collection_modifyitems(items, config):
-    include_no_vnodes_tests = not dtest_config.use_vnodes
-    include_no_offheap_memtables_tests = not dtest_config.use_off_heap_memtables
+    if skip_conditions.skip_resource_intensive_tests:
+        logger.info("Resource intensive tests will be skipped because "
+                    "it was requested to skip or there is not enough system resource "
+                    "and --force-resource-intensive-tests or --only-resource-intensive-tests "
+                    "were not specified")
```
I don't see why skipping resource intensive tests with not enough resources and `--only-resource-intensive-tests` is wrong. IMO it makes sense to require the `--force-resource-intensive-tests` flag in that case, since one flag controls the selection and the other one skips the resource check. Let's see what the second reviewer thinks about this. In any case if we are going to change the behaviour of the flags or we find them confusing we should probably extend their descriptions in `pytest_addoption` and `run_dtests.py`.

k-rus replied:
I explained my thoughts in the Jira ticket, so I guess we continue there and let's see what others think about it.

k-rus committed:
Address reviewer comments

k-rus committed:
Fix formatting check bug in using flake8

flake8 contains two warnings W503 and W504, which contradict each
other on Python code if code contains binary operators such as OR or
AND and is split into multiple lines:

W503 line break before binary operator
W504 line break after binary operator

The recommended code style is to place line break after a binary
operator. Thus this fix adds W503 to list of flake8 errors and
warnings to ignore.

Also adding .pytest_cache/ to .gitignore

k-rus committed:
Fix unexpectedly ignored dtests

This commit fixes few unexpected behaviors, which were:
1. If any test class is marked with `resource_intensive`, then all tests
   in the file were treated as resource intensive.
2. If one test class is marked with `vnodes` and another test class is
   marked with `no_vnodes` in the same file, then all tests were
   skipped in the file.

The unexpected behaviors are due to marking all tests in a file if any
class is assigned the mark. This commit fixes that marks on other
classes in the file do not affect a test, which is not marked and
neither its class. This is also applied to upgrade tests, so the code
is simplified and behavior is unified.

Since the upgrade tests were utilizing that the marking one class is
propagated to all classes in the file, they were revisited. In many
cases the upgrade tests were already included, because their classes
inherited from a class, which already was marked. In such cases
unnecessary marks were removed.

patch by Ruslan Fomkin; reviewed by Andrés de la Peña and Ekaterina
Dimitrova for CASSANDRA-16841

k-rus commented:
I addressed comments and started [new build in CircleCI](https://app.circleci.com/pipelines/github/k-rus/cassandra?branch=CASS-16841-CI).

k-rus committed:
Require force execution of resource intensive

k-rus commented:
Addressed comment and run [j11_dtest build](https://app.circleci.com/pipelines/github/k-rus/cassandra/10/workflows/e56bdedd-45ba-466f-ae32-a003b9004063)

adelapena commented on conftest.py:
```diff
@@ -605,28 +603,16 @@ def pytest_collection_modifyitems(items, config):
+    if skip_conditions.skip_resource_intensive_tests:
+        logger.info("Resource intensive tests will be skipped because "
+                    "it was requested to skip or there is not enough system resource "
+                    "and --force-resource-intensive-tests were not specified")
```
```suggestion
                    "and --force-resource-intensive-tests was not specified")
```

k-rus replied:
Thank you for catching this!

k-rus committed:
Fix message about skipping resource intensive tests

k-rus committed:
Fix duplicate execution of imported tests

Some upgrade tests import and then inherit base classes from
non-upgrade folder. Since the base classes were following PyTest
notation to be picked up for the execution, the same tests were
executed two times. This commit removes duplicate execution by
renaming the base classes to be not picked up by PyTest. In one case
it was necessary to introduce a child class, which follows PyTest
convention and inherits from the base class in the same non-upgrade
file.

adelapena commented on bootstrap_test.py:
```diff
@@ -1097,3 +1097,12 @@ def test_host_id_override(self):
         session1 = self.patient_exclusive_cql_connection(node1)
         assert_one(session1, "SELECT host_id FROM system.peers_v2 WHERE peer = {}".format(address2), [uuid.UUID(host_id)])
+
+class TestBootstrap(BootstrapTester):
```
Nit: missed additional blank line before the class declaration

k-rus replied:
Thank you for noting! I haven't set up automatic formatting on commits and there is nothing prepared in Cassandra to just reuse. I know that there is git commit hook, but it doesn't fix formatting, it is just erroring.

And took almost a day to get Travis to run it.

I will fix it soon.

ekaterinadimitrova2 replied:
I use PyCharm and I get warnings for formatting, never had to do anything in addition to be able to see those.

k-rus replied:
I use VS Code. My preferable approach is to be not dependent on IDE, thus I am looking into setting up a git commit hook.
